### PR TITLE
Bazel regen rule to generate cubemx code

### PR DESCRIPTION
**Pull request by akhilveeraghanta**
### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/firmware_new/tools/regen.sh
Comment (garethellis0) [inline src/firmware_new/tools/regen.sh]: do we actually need this bazel stuff? This gives us the ability to do stuff like `$(location //software/geom:util)`), but I don't see anything like that here?

File: src/firmware_new/tools/regen.sh
Comment (garethellis0) [inline src/firmware_new/tools/regen.sh]:
``` suggestion
    echo "Usage: $THIS_SCRIPT_FILENAME path/to/folder/with/.ioc/file/"
```

File: src/firmware_new/tools/regen.sh
Comment (garethellis0) [inline src/firmware_new/tools/regen.sh]: I would perhaps advise against nuking these folders, I think it should be OK if there are extra files in them, but I don't want to risk deleting peoples work.

File: src/firmware_new/tools/regen.sh
Comment (garethellis0) [inline src/firmware_new/tools/regen.sh]: We should add "Cube" folders (like `Src` and `Inc`) to the `.gitignore` so we don't accidentally commit these.

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Looking good, just a few small comments!

File: src/firmware_new/tools/regen.sh
Comment (akhilveeraghanta) [reply src/firmware_new/tools/regen.sh]: yeah I was using it earlier this week but forgot to nuke it today, removed

File: src/firmware_new/tools/regen.sh
Comment (akhilveeraghanta) [reply src/firmware_new/tools/regen.sh]: fixed

File: src/firmware_new/tools/regen.sh
Comment (akhilveeraghanta) [reply src/firmware_new/tools/regen.sh]: These are in the `/tmp` directory, but I removed the `rm -rf` just to be safe

File: src/firmware_new/tools/regen.sh
Comment (akhilveeraghanta) [reply src/firmware_new/tools/regen.sh]: These are in the `/tmp/` dir, the bazel rule copies them there, regen's and copies them back

**Review by henryymliu — CHANGES_REQUESTED** [review]
looks good overall

File: src/firmware_new/tools/regen.sh
Comment (garethellis0) [reply src/firmware_new/tools/regen.sh]: OH! That's totally fine then, whatever you think is best here.

**Review by garethellis0 — DISMISSED** [review]
LGTM (pending Henry's comments) :+1: :tada: - feel free to re-add the `rm -rf` or not, totally your call since it's in the `/tmp` directory

File: environment_setup/install_cubemx.sh
Comment (akhilveeraghanta) [reply environment_setup/install_cubemx.sh]: fixed

File: src/firmware_new/tools/regen.sh
Comment (akhilveeraghanta) [reply src/firmware_new/tools/regen.sh]: fixed

**Review by henryymliu — DISMISSED** [review]
LGTM

File: src/firmware_new/tools/regen.sh
Comment (MathewMacDougall) [inline src/firmware_new/tools/regen.sh]:
```suggestion
# Generate code in the project directory
```

File: src/firmware_new/tools/BUILD
Comment (MathewMacDougall) [inline src/firmware_new/tools/BUILD]: nit: Can we name this something slightly more specific, like `regen_cubemx`? (Same goes for the `regen.sh` file)

Running a target called `//firmware_new/tools:regen` doesn't really let me know what it's doing.

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
2 nits, otherwise LGTM

File: src/firmware_new/tools/regen.sh
Comment (akhilveeraghanta) [reply src/firmware_new/tools/regen.sh]: fixed

File: src/firmware_new/tools/BUILD
Comment (akhilveeraghanta) [reply src/firmware_new/tools/BUILD]: yeah agreed, changed to `cubemx_regen`

**Review by MathewMacDougall — DISMISSED** [review]
LGTM :+1:

**Review by henryymliu — DISMISSED** [review]
🚢🤡

**Review by garethellis0 — APPROVED** [review]

**Comment by akhilveeraghanta** [comment]
Could one of you guys try this on your computers and make sure 1) it installs correctly and 2) the cubemx output looks good?

I noticed its spewing out a bunch of "invalid parameter" log msgs, not sure if that's normal

**Review by garethellis0 — CHANGES_REQUESTED** [review]
A few things after playing with this locally:
- we should make sure we put a symlink in like `/usr/local/bin/STM32CubeMX`->`/opt/STM32CubeMX/STM32CubeMX` so that the elec folks can easily launch and use the tool to do the actual pin-mapping
- we should version our installs of Cube, so that we don't clobber stuff in the future when we update versions, and we keep the old versions around just in case. See how `install_clion.sh` does this (by creating different folders in `/opt` and just symlinking to the newest one)
- generated firmware looks fine! flashed it to the board and was able to debug with no issues

Good call asking for extra testing! I'll try to make a habit of doing this anyhow for future similar PR's

**Review by MathewMacDougall — APPROVED** [review]
LGTM :+1:

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada:

**Review by henryymliu — APPROVED** [review]
LGTM, looks like Travis got stuck so you may want to push an empty commit or something

**Comment by garethellis0** [comment]
Funky, looks like Travis passed as well? I've restarted it, but might need another commit if that doesn't work.

**Comment by MathewMacDougall** [comment]
Looks good now.

**Pull request merged**
